Source/DafnyCore/AST/DafnyAst.cs

```diff
@@ -31,6 +33,15 @@ public interface INode {
     /// losing source location information, so those transformed nodes should not be returned by this property.
     /// </summary>
     IEnumerable<INode> Children { get; }
+
+    /// <summary>
+    /// These children should match what was parsed before the resolution phase.
+    /// That way, gathering all OwnedTokens of all recursive ConcreteChildren should result in a comprehensive
+    /// coverage of the original program
+    /// </summary>
+    IEnumerable<INode> ConcreteChildren { get; }
```

Comment:
I feel good about this idea ! However, that might require some architecturing outside of the scope of this PR, because the language server does not provide the document after parsing, only after full resolution.

Reply:
I think it is in scope, because this PR adds code because that feature is missing. 

Also, I also don't think we should expose the Dafny `Program` to the language server handlers, so we need a formatting specific change to `Compilation`, like what I described in `https://github.com/dafny-lang/dafny/pull/2399#discussion_r1022771435`.

Reply:
> Also, aren't changes that allow us to write nicer PR code always in scope?

I wholeheartedly agree. But by architectural decisions, I really mean answering two not trivial questions:
- Do we really want to have SubExpressions to contain non-resolved sub expressions after parsing, and resolved sub-expressions after resolution? These are two different set of expressions. A developer might be surprised to see Dafny recurse in one part of the tree when they expected the other part (sometimes they are mutually exclusive, like for `ConcreteSyntaxExpression`)
- The Language Server wants to access both pre-resolved expressions (for getting all the 1-to-1 display in the program) and sometimes post-resolved expressions (for having the correct types). We already have work-arounds in place to traverse the tree of expressions pre-resolution in the language server. Don't we want this traversal directly in the tree? Do we have evidence that it can work on two different programs independently?
- Don't we want to instead put every expression into a ConcreteSyntaxExpression?
- Ideally, DafnyLanguageServer depends on DafnyCore, not the way around. So the change for the DafnyLanguageServer should follow the formatter, not constrain it. It would make perfect sense to investigate how to split the resolved document in two phases in a subsequent PR and at the same time merge the two notions of children, rather than trying to merge two different notions of children and changing the language server right now in this PR.

If the answer was obvious, then I would do the change right away. It is not obvious to me and will lead to many discussions which I think will vastly delay the merging of this PR. I'm happy to postpone it indefinitely until we resolve these issues, but not everyone might be happy about that.

If you are worried about exposing a new API to remove it later, then we might want to mark the field PreResolveChildren as "internal", [even if it is in an interface](https://stackoverflow.com/a/368482/1287856).

Reply:
> Do we really want to have SubExpressions to contain non-resolved sub expressions after parsing, and resolved sub-expressions after resolution?

I think so, since this means the AST after resolution behaves as if it only has its resolved state, and no longer the pre-resolved state. I believe this 'single state AST' is what we want to work towards. Even if the internal state of the AST nodes can still be 'two-state', it's already nice if the APIs behave single-state.

> The Language Server wants to access both pre-resolved expressions (for getting all the 1-to-1 display in the program) and sometimes post-resolved expressions (for having the correct types). 

I think that's easy to achieve by cloning the AST after parsing, so I don't feel that needs to influence any design.

> Don't we want to instead put every expression into a ConcreteSyntaxExpression?

I'm hoping we can get rid of `ConcreteSyntaxExpression`

> It would make perfect sense to investigate how to split the resolved document in two phases in a subsequent PR and at the same time merge the two notions of children, rather than trying to merge two different notions of children and changing the language server right now in this PR.

My guess is that doing it now results in significantly fewer changed lines of code in this PR. Exposing the post-parsing document in the language server seems simple to me, and a single Children that supports both pre and post-resolve seems smaller than having two separate ones.

Reply:
> > Do we really want to have SubExpressions to contain non-resolved sub expressions after parsing, and resolved sub-expressions after resolution?
> 
> I think so, since this means the AST after resolution behaves as if it only has its resolved state, and no longer the pre-resolved state. I believe this 'single state AST' is what we want to work towards. Even if the internal state of the AST nodes can still be 'two-state', it's already nice if the APIs behave single-state.
> 
> > The Language Server wants to access both pre-resolved expressions (for getting all the 1-to-1 display in the program) and sometimes post-resolved expressions (for having the correct types).
> 
> I think that's easy to achieve by cloning the AST after parsing, so I don't feel that needs to influence any design.
> I'm hoping we can get rid of `ConcreteSyntaxExpression`

Cloning the AST after parsing has its own pros and cons. I'm inclined to actually merge the two notions of expressions and do what you suggest, but please consider the following:

Pros:
* Faster time to activate the formatter in the IDE, no need to resolve the syntax tree.

Cons:
* It creates an AST that is guaranteed to not have any typing or name resolution information. If, at some point, the printer/formatter wanted to access the resolution data (e.g. types, name resolution) to format things in an enhanced way (e.g. making variable types at declaration explicit, synchronizing formal types in the comments

> > It would make perfect sense to investigate how to split the resolved document in two phases in a subsequent PR and at the same time merge the two notions of children, rather than trying to merge two different notions of children and changing the language server right now in this PR.
> 
> My guess is that doing it now results in significantly fewer changed lines of code in this PR. Exposing the post-parsing document in the language server seems simple to me, and a single Children that supports both pre and post-resolve seems smaller than having two separate ones.

Ok, if you are ready to review this extra change, then I can do it.
For now, let's focus on making https://github.com/dafny-lang/dafny/pull/3070 acceptable.

Reply:
That's an excellent idea.